### PR TITLE
Color override

### LINKED ISSUE #868: What is the new way of using getFillColor in Chart options
- [ ] Feature request
- [ ] Design defect
- [x] Source code defect
- [ ] Demo/documentation defect
- [ ] Other

I'm using `@carbon/charts-react`. There is a breaking change between 0.40.2 and 0.40.3 about the behaviour of [`getFillColor`](https://carbon-design-system.github.io/carbon-charts/documentation/interfaces/_interfaces_charts_.basechartoptions.html#getfillcolor)

## v0.40.2 

In version 0.40.2 and below, the following code works to set the colors in a Donut Chart 

```javascript
import React from "react";
import ReactDOM from "react-dom";
import { DonutChart } from "@carbon/charts-react";
import "@carbon/charts/styles.css";

const colors = {
  java: "#FF0000",
  javascript: "#00FF00",
  "c++": "#0000FF"
};

const chartData = [
  {
    group: "java",
    value: 300
  },
  {
    group: "javascript",
    value: 600
  },
  {
    group: "c++",
    value: 200
  }
];

const demoDonutOptions = {
  getFillColor: (field) => {
    return colors[field];
  },
  height: "300px"
};

function App() {
  return (
    <div className="App">
      <h3>React Donut chart with label</h3>
      <div>
        <DonutChart data={chartData} options={demoDonutOptions} />
      </div>
    </div>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```
See [this sandbox](https://codesandbox.io/s/carbon-react-donut-color-0402-0u03z?file=/src/index.js)

It renders the colors correctly (calling getFillColor) :

chart-color

## v0.40.3 and above

It doesn't work anymore, see [this sandbox](https://codesandbox.io/s/carbon-react-donut-color-04013-1lq2v?file=/src/index.js) which uses the latest version at this time (0.40.13)

It doesn't set the color anymore, the `getFillColor` function is not called, the default colors are rendered :

chart-color2

Maybe there are some option fields missing in order for it to work in 0.40.3 and above ? When I look at [the history](https://github.com/carbon-design-system/carbon-charts/commit/8e32712652a4726da5a978313e01c1139d8baf77) I've noticed some conditional change when `getFillColor` is called related to `color.scale` field. Maybe @sophiiae or @theiliad have more insights about the changes in 0.40.3

Thanks a lot for your help

### PR DESCRIPTION
**Pull request by sophiiae**
user can partially override chart color.
e.g., provide color scale for only 2 data groups when the chart involves 3 or more data groups

https://deploy-preview-896--carbon-charts-core.netlify.app/?path=/story/bar-vertical--simple-bar-color-override

also closes  #868

### CODE REVIEW

File: packages/core/src/model.ts [hunk 1]
```diff
@@ -483,18 +484,17 @@ export class ChartModel {
 		}
 
 		const options = this.getOptions();
-		const defaultStrokeColor = this.colorScale(group);
+		const defaultStrokeColor = this.colorScale[group]
```
Comment (theiliad) [inline packages/core/src/model.ts:487]: just switch this to `Tools.getProperty()`

File: packages/core/src/components/graphs/treemap.ts [hunk 2]
```diff
@@ -336,7 +346,12 @@ export class Treemap extends Component {
 							"graph_element_mouseout_fill_update"
 						)
 					)
-					.style("fill", null);
+					.style("fill", (d: any) => {
+						const fillColor = self.model.getFillColor(
+							d.parent.data.name
+						);
+						return fillColor ? fillColor : null;
```
Comment (theiliad) [inline packages/core/src/components/graphs/treemap.ts:353]: wouldn't `getFillColor` return null anyways if it doesn't like the input provided?

**Review by theiliad — CHANGES_REQUESTED** [review]

File: packages/core/src/components/graphs/treemap.ts (hunk 2 above)
Comment (sophiiae) [reply packages/core/src/components/graphs/treemap.ts:353]: it returns `undefined` if no provide color

Comment (theiliad) [reply packages/core/src/components/graphs/treemap.ts:353]: it should return null...?

Comment (theiliad) [reply packages/core/src/components/graphs/treemap.ts:353]: it should return null

**Review by theiliad — CHANGES_REQUESTED** [review]

**Review by theiliad — APPROVED** [review]

**Review by natashadecoste — APPROVED** [review]
💯 looks good

File: packages/core/src/components/graphs/gauge.ts [hunk 3]
```diff
@@ -359,7 +365,7 @@ export class Gauge extends Component {
 				"class",
 				status !== null ? `gauge-delta-arrow status--${status}` : ""
 			)
-			.attr("fill", () => (status === null ? "currentColor" : null))
+			.style("fill", () => (status === null ? "currentColor" : null))
```
Comment (theiliad) [inline packages/core/src/components/graphs/gauge.ts:368]: what's `currentColor` do?

File: packages/core/src/components/graphs/gauge.ts [hunk 4]
```diff
@@ -140,7 +146,7 @@ export class Gauge extends Component {
 					originalClassName: "arc-foreground"
 				})
 			)
-			.attr("fill", (d) => self.model.getFillColor(d[groupMapsTo]))
+			.style("fill", (d) => (customColor ? customColor : null))
```
Comment (theiliad) [inline packages/core/src/components/graphs/gauge.ts:149]: `Tools.getProperty` returns null anyways if the path doesn't exist

File: packages/core/src/components/essentials/legend.ts [hunk 5]
```diff
@@ -95,20 +95,18 @@ export class Legend extends Component {
...
+			.attr("class", (d, i) =>
+				this.model.getColorClassName({
+					classNameTypes: [ColorClassNameTypes.FILL],
+					dataGroupName: d.name,
+					originalClassName: "checkbox"
+				})
+			)
+			.style("fill", (d) => (
+				d.status === Configuration.legend.items.status.ACTIVE
+					? this.model.getFillColor(d.name) || this.model.getStrokeColor(d.name)
```
Comment (theiliad) [inline packages/core/src/components/essentials/legend.ts:107]: don't `getFillColor` & `getStrokeColor` return null anyways? `null || null` is `null`, we don't need the ternary

**Review by theiliad — CHANGES_REQUESTED** [review]

File: packages/core/src/components/graphs/gauge.ts (hunk 3 above)
Comment (sophiiae) [reply packages/core/src/components/graphs/gauge.ts:368]: I'm not sure, I haven't found the implicit connection between `currentColor` and other stuff. I didn't touch this part in this PR.

File: packages/core/src/components/essentials/legend.ts (hunk 5 above)
Comment (sophiiae) [reply packages/core/src/components/essentials/legend.ts:107]: I tested the `getCustomStroke` that user can custom `line chart` color, and the legend remain the default color palette. That's why I add the ternary.

File: packages/core/src/components/graphs/gauge.ts (hunk 3 above)
Comment (sophiiae) [reply packages/core/src/components/graphs/gauge.ts]: I removed this line since it doesn't make any difference and the color is assigned by the classname.

Comment (theiliad) [reply packages/core/src/components/graphs/gauge.ts]: could you remove the line and test whether the graph still fully functions?

Comment (sophiiae) [reply packages/core/src/components/graphs/gauge.ts]: I did in 1a11c07

**Review by theiliad — APPROVED** [review]

**Review by natashadecoste — APPROVED** [review]

**Pull request merged**